modules/howtos/examples/Import.java

```diff
@@ -0,0 +1,411 @@
+    // end::connect[]
+    
+    // tag::reactiveCollection[]
+    reactiveCollection = collection.reactive();
+    // end::reactiveCollection[]
+  }
+  
+  public static void main(String... args) {
+    Import importer = new Import();
+    
+    importer.importCSV();
+    importer.importTSV();
+    importer.importJSON();
+    importer.importJSONL();
+    importer.importCSV_batch();
+    importer.importTSV_batch();
+    importer.importJSON_batch();
+    importer.importJSONL_batch();
+  }
+  
+  // NOTE: non-generic Map, because the readers produce slightly different outputs:
+  //   CSV: Map<String,String>
+  //   JSON: Map<String,Object>
+  // and as generics annotations are lost at runtime we can't have
+  // a single overloaded method to deal with those.
+  //
+  // In real code, you might choose to cast from Object to Map<String,Object>
+  // or define multiple methods, but those confuse the example a little.
+  
+  // tag::preprocess[]
+  public JsonDocument preprocess(Map row) {
+    Map value = new HashMap(row);
+
+    // define the KEY
+    String key = value.get("type") + "_" + value.get("id");
+
+    // do any additional processing
+    value.put("importer", "Java SDK");
+    
+    return new JsonDocument(key, JsonObject.from(value));
+  }
+  // end::preprocess[]
+  
+  // tag::upsertDocument[]
+  public void upsertRow(Map row) {
+    
+    JsonDocument doc = preprocess(row);
+    
+    String key = doc.getId();
+    Object value = doc.getContent();
+    
+    // upsert the document
+    collection.upsert(key, value);
+    
+    // any required logging
+    System.out.println(key);
+    System.out.println(value);
+  }
+  // end::upsertDocument[]
+  
+  
+  // tag::importCSV[]
+  public void importCSV() {
+    try {
+      CSVReaderHeaderAware csv = new CSVReaderHeaderAware(
+        new FileReader("howtos/examples/import.csv"));
+      
+      Map<String, String> row;
+      while ((row = csv.readMap()) != null) {
+        upsertRow(row);
+      }
+    }
+    catch (java.io.FileNotFoundException e) {
+      System.out.println("handle FileNotFoundException...");
+    }
+    catch (java.io.IOException e) {
+      System.out.println("handle IOException...");
+    }
+    catch (com.opencsv.exceptions.CsvValidationException e) {
+      System.out.println("handle CsvValidationException...");
+    }
+  }
+  // end::importCSV[]
+  
+  
+  // tag::importCSV_batch[]
+  public void importCSV_batch() {
+    // tag::omit[]
+    System.out.println("importCSV_batch");
+    // end::omit[]
+    
+    Flux<Map<String,String>> rows = Flux.generate(
+      
+      () -> new CSVReaderHeaderAware(
+        new FileReader("howtos/examples/import.csv")),
+      
+      (state, sink) -> {
+        try {
+          Map<String,String> row = state.readMap();
+          if (row == null) { sink.complete(); }
+          else { sink.next(row); }
+          return state;
+        }
+        catch (IOException e) { throw new Error("IOException"); }
```

Comment:
Instead of catching different exceptions and re-throwing, how about signalling the error like:

    catch (Exception t) {
        sink.error(t);
    }

As an aside, `Error` is typically reserved for conditions that should never, *EVER* happen. Like, inconceivable situations that indicate something is fundamentally wrong with the code. If you just want to smuggle a checked exception, wrapping it in a `RuntimeException` is usually the way to go.

Reply:
good idea! (I wasn't sure if I should name each exception to document which ones might happen? But happy to keep it simpler...)

Reply:
hmm, sink.error gives me a Generics type error, but I've used `throw new RuntimeException(e)` elsewhere at your suggestion anyway, so may keep with that for consistency.